gmso/formats/lammpsdata.py

```diff
@@ -656,11 +768,25 @@ def _write_header(out_file, top, atom_style):
             "{:d} angle types\n".format(len(top.angle_types(filter_by=pfilter)))
         )
     if top.n_dihedrals > 0 and atom_style in ["full", "molecular"]:
-        out_file.write(
-            "{:d} dihedral types\n".format(
-                len(top.dihedral_types(filter_by=pfilter))
+        unique_dtypes = top.dihedral_types(filter_by=pfilter)
+
+        nkeys = len(next(iter(unique_dtypes)).parameters.keys())
+        nparams = len(
+            list(
+                _flatten(
+                    (
+                        map(
+                            lambda x: [
+                                val.tolist() for val in x.parameters.values()
+                            ],
+                            unique_dtypes,
+                        )
+                    )
+                )
             )
         )
+        ntypes = int(nparams / nkeys)
```

Comment:
I am not super sure what this block is doing, I will loop back to this later.

Reply:
Finally got what this piece is doing. I think the logic is sounds, but the code itself is not straightforward and may pose some issue for maintenance in the future. I get it is due to the `TopologyPotentialView` object is set up, but I think we should add a few lines explaining what each var is trying to do here.

Reply:
fixed